src/github/utils.ts

```diff
@@ -1110,13 +1109,98 @@ export function hasEnterpriseUri(): boolean {
 	return !!getEnterpriseUri();
 }
 
-export function generateGravatarUrl(gravatarId: string | undefined, size: number = 200): string | undefined {
-	return !!gravatarId ? `https://www.gravatar.com/avatar/${gravatarId}?s=${size}&d=retro` : undefined;
+function generateGravatarUrl(gravatarId: string | undefined, size: number = 200): string | undefined {
+	if (!gravatarId) {
+		return undefined;
+	}
+
+	// This URL is a publicly available render of the `resources/icons/dark/github.svg` as a 2048x2048 png
+	const invertocatEncodedUrl = encodeURIComponent('https://user-images.githubusercontent.com/675956/216480776-611f6c3a-6334-46d3-8f5d-9a11ae62df6d.png');
+	const invertocatStyle = 'invertocat';
+
+	let style = vscode.workspace.getConfiguration('githubPullRequests').get<string>('defaultGravatarsStyle', invertocatStyle);
+	if (style === invertocatStyle) {
+		style = invertocatEncodedUrl;
+	} else if (style === 'none') {
+		return undefined;
+	}
+
+	return `https://www.gravatar.com/avatar/${gravatarId}?s=${size}&d=${style}`;
 }
 
-export function getAvatarWithEnterpriseFallback(avatarUrl: string, email: string | undefined, authProviderId: AuthProvider): string | undefined {
-	return authProviderId === AuthProvider.github ? avatarUrl : (email ? generateGravatarUrl(
-		crypto.createHash('md5').update(email?.trim()?.toLowerCase()).digest('hex')) : undefined);
+const enterpriseAvatarQueue = new PQueue({concurrency: 3});
+const enterpriseAvatarCache: {[k: string]: Promise<string | void>} = {};
+
+async function getEnterpriseAvatarUrl(avatarUrl: string | undefined, octokit: LoggingOctokit): Promise<string | undefined> {
+	try {
+		if (!avatarUrl) {
+			return;
+		}
+
+		const avatarUri = vscode.Uri.parse(avatarUrl, true);
+		const enterpriseUri = getEnterpriseUri()!;
+		const enterpriseAvatarRestBase = '/enterprise/avatars';
+
+		// static asset from enterprise does not need authentication
+		if (avatarUri.scheme === 'data' || avatarUri.authority === `assets.${enterpriseUri.authority}`) {
+			return avatarUrl;
+		}
+
+		if (avatarUri.authority !== `avatars.${enterpriseUri.authority}`) {
+			return;
+		}
+
+		const cacheKey = `${avatarUri.path}?${avatarUri.query}`;
+		const options = {};
+		const qs = new URLSearchParams(avatarUri.query);
+
+		qs.forEach((v, k) => {
+			options[k] = v;
+		});
+
+		if (!(cacheKey in enterpriseAvatarCache)) {
+			enterpriseAvatarCache[cacheKey] = enterpriseAvatarQueue.add(() =>
+				octokit.api.request(`GET ${enterpriseAvatarRestBase}${avatarUri.path}`, options).then(
+					resp => {
+						const dataUri = `data:${resp.headers['content-type']};base64,${Buffer.from(resp.data).toString(
+							'base64',
+						)}`;
+						return dataUri;
+					},
+					() => {
+						delete enterpriseAvatarCache[cacheKey];
+						return;
+					},
+				),
```

Comment:
Why is this needed?

Reply:
See https://github.com/microsoft/vscode-pull-request-github/pull/4497#issuecomment-1423080477